### ISSUE TITLE
Spurious warning when cloning `Copy` member of packed struct

### ISSUE THREAD

**Issue by hniksic** [rust-lang/rust#70019]
I expected the following code to compile without warning:

```rust
#[repr(C,packed)]
struct Foo {
    memb: u64,
}

fn main() {
    let f = Foo { memb: 0 };
    println!("{}",  f.memb.clone());
}
```

However, with Rust 1.41 it warns:

```
warning: borrow of packed field is unsafe and requires unsafe function or block (error E0133)
 --> src/main.rs:8:21
  |
8 |     println!("{}",  f.memb.clone());
  |                     ^^^^^^
  |
  = note: `#[warn(safe_packed_borrows)]` on by default
```

Without `clone()` the warning is correct because `println!()` takes the a reference to the packed value and passes it to the formatting mechanism, which can cause unaligned access. However, I would expect `clone()` to copy the underlying value  and pass that to `println!`.

I am aware that the warning comes from `clone` itself, because it technically receives a reference. But I would expect the implementation of `Clone` to work on packed values that are `Copy`, because copying them is already supported and safe. For example, replacing `f.memb.clone()` with `f.memb + 0` eliminates the warning.

In summary, `x.clone()` on a packed `u64` should be as safe as `x + 0`, and I would expect it behave the same.

**Comment by jonas-schievink** [comment]
I think the problem here is that `clone` takes `&self`, and that reference might be unaligned, which is unsafe.

**Comment by hniksic** [comment]
@jonas-schievink That is acknowledged in the second-to-last paragraph of the description. The`clone()` implementation is not supplied by the user, it is provided automatically. Since Rust knows how to copy the value safely and without warning, shouldn't it be able to clone it the same way?

**Comment by jonas-schievink** [comment]
If you call `clone`, you implicitly create a reference to the value, and then it's already too late since the reference might not be properly aligned, which is UB

**Comment by hniksic** [comment]
That is the case for user-provided functions, but `u64::clone` is an intrinsic provided by the compiler. For `Copy` types, `Clone` simply copies the value, so cloning is exactly as safe as copying.

**Comment by sfackler** [comment]
> That is the case for user-provided functions, but u64::clone is an intrinsic provided by the compiler.

No, it is not: https://doc.rust-lang.org/src/core/clone.rs.html#181-186

**Comment by hniksic** [comment]
@sfackler I stand corrected, but note that `clone` is still provided by Rust, not by the user. The implementation does the same kind of copying that `foo.memb + 0` would do, right?

**Comment by qnighy** [comment]
Your code is equivalent to `Clone::clone(&f.memb)`. The problem is that even `let x = &f.memb;` is UB.

**Comment by Centril** [comment]
Closing as not-a-bug.
